### ISSUE TITLE
macos platform channels can't distinguish `nil` and empty data

### ISSUE THREAD

**Issue by gaaclarke** [flutter/flutter#110865]
The following platform channel integration tests fail on macos:
- https://github.com/flutter/flutter/blob/f57fcac717092e9d55d25a63d79505b455719609/dev/integration_tests/channels/lib/main.dart#L113:L113
- https://github.com/flutter/flutter/blob/f57fcac717092e9d55d25a63d79505b455719609/dev/integration_tests/channels/lib/main.dart#L117:L117

I spent some time debugging it and it ended up being a bit of an investment tracking down all the places we throw away the distinction between `nil` and empty data.

**Comment by stuartmorgan** [comment]
This was deliberate per https://github.com/flutter/engine/pull/25064; we should probably just adjust the tests accordingly.

**Comment by gaaclarke** [comment]
> This was deliberate per [flutter/engine#25064](https://github.com/flutter/engine/pull/25064); we should probably just adjust the tests accordingly.

Are you suggesting the test should test one thing if it's iOS and another thing if it's macOS?  Or are you suggesting that we change iOS behavior to match macOS behavior?  I'm not keen on either of those options.  iOS conventions are more established and if anything the macOS implementation should be changed to match iOS.

**Comment by stuartmorgan** [comment]
> Are you suggesting the test should test one thing if it's iOS and another thing if it's macOS?

Yes, since we made a deliberate decision to have this behavior, and our tests should reflect that.

> iOS conventions are more established

Not for macOS plugins, which are the consumers of that API.

> and if anything the macOS implementation should be changed to match iOS.

That would potentially break existing, published plugins.

**Comment by gaaclarke** [comment]
> That would potentially break existing, published plugins.

It's not clear to me that the cost of breaking them is less than the cost of maintaining 2 separate semantics.  Especially when it's easy to share code between iOS and macOS.  Practically speaking I don't think this will show up often though.

For now I'll convert the comments in that integration test from TODO's to notes about the difference.  I suggest we keep the issue open to see if we get more data to inform the prudence of this decision.

**Comment by cbracken** [comment]
Oh interesting -- I'd forgotten it was me who introduced the change that fixed the crash but also beahviour that differs from iOS. Agreed it would be advantageous for us to make them consistent. Will see if I can remember why we elected to go with different behaviour (especially given that my comment suggests I was attempting to be consistent with iOS).

**Comment by stuartmorgan** [comment]
> Practically speaking I don't think this will show up often though.

Since it took several years for the earlier iteration of this--which was a *crash*--to be reported, that's a safe bet.



> I suggest we keep the issue open to see if we get more data to inform the prudence of this decision.

We can, but we already have a year and a half of issue tracker data on this (namely, that it's never been reported), and waiting longer only increases the chance of breaking existing code.

Ideally it would have matched from the start, but since it has had different behavior for quite a while I think we should consider the ship as having sailed. (IIRC matching the iOS behavior was tricky for some reason?)

**Comment by cbracken** [comment]
I also remember it being tricky to match the iOS behaviour exactly but I don't remember the details -- it's been quite a while. Tagging as `p5`.
